build: support native cross-compilation for ARM64

sarandia opened the pull request:
Summary of Changes
1. A new cmake file, ./cmake/linux-arm64.cmake, to configure the build
to use the aarch64-linux-gnu toolchain.
2. Changes in ./CMakeLists.txt to update the CONFIGURE and BUILD
commands for libbacktrace and jemalloc.
3. Changes in ./cmake/luajit.cmake and ./cmake/onigmo.cmake to update
the CONFIGURE and BUILD commands for luajit and onigmo libraries.

Benefits
1. The build time using native cross-compilation for arm64 is much
faster than that using QEMU userspace emulation, which is the current
method used by OSS to release multi-arch images. In our experiments, it
was ~20X faster (~1min vs. ~20min).
2. Supporting cross-platform builds natively will allow users to set up
builds for multiple architectures in their build pipelines without
supplying additional configurations or Dockerfiles.
3. We will be able to easily support additional architectures (e.g.
armv7) cross-compilation by setting the right toolchain path during
cmake.

https://github.com/fluent/fluent-bit/issues/3119
----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [N/A] Example configuration file for the change
- [N/A] Debug log output from testing the change
<!-- Invoke Fluent Bit and Valgrind as: $ valgrind ./bin/fluent-bit <args> -->
- [N/A] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

**Documentation**
<!-- Docs can be edited at https://github.com/fluent/fluent-bit-docs -->
- [X] Documentation required for this feature

<!--  Doc PR (not required but highly recommended) -->

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

niedbalski commented:
Please check https://github.com/fluent/fluent-bit/pull/3151 before merging this. Thanks

edsiper commented:
#3151 has been merged.

@niedbalski can you review this PR ?

sarandia commented:
@niedbalski  I rebased this PR and moved our new Dockerfile to dockerfiles/.

niedbalski commented on cmake/luajit.cmake:
```diff
@@ -22,7 +22,7 @@ ExternalProject_Add(luajit
   SOURCE_DIR ${LUAJIT_SRC}
   CONFIGURE_COMMAND ./configure
-  BUILD_COMMAND $(MAKE) CC=${CMAKE_C_COMPILER} CFLAGS=${CFLAGS} BUILD_MODE=static "XCFLAGS=-fPIC"
+  BUILD_COMMAND $(MAKE) CROSS=${CROSS_PREFIX} CFLAGS=${CFLAGS} BUILD_MODE=static "XCFLAGS=-fPIC"
```
what is the CROSS_PREFIX variable set to on non arm64 targets?

sarandia replied:
https://github.com/fluent/fluent-bit/blob/92e5e9f33f011da1ec94737dce4b31954556f165/lib/luajit-2.1.0-1e66d0f/doc/install.html#L377

It is a prefix for the path to the toolchain. It is empty for X86 builds (if the compilation is done on an X86 host machine), and luajit will just use "" + default_path for the toolchain. For other architectures, we need dedicated cmake files for each architecture to set this accordingly.

niedbalski commented on dockerfiles/Dockerfile.arm64v8-dockerx:
```diff
@@ -0,0 +1,79 @@
+# Get arm64 dependency
```
There is already a dockerfiles/Dockerfile.arm64v8 file, In my opinion, if we are going to get rid off qemu binary emulation, we should then use that file and adapt it instead.

niedbalski replied:
@sarandia another change I'd like to address on this proposal is to add dockerfile/Dockerfile.armv8 dockerfile/Dockerfile.armv8-master and dockerfile/Dockerfile.armv8-debug in the same line of amd64.

sarandia replied:
The DFLB flags are the same in Dockerfile.arm64v8-buildx and Dockerfile.arm64v8, so they should produce the same image.

Can we submit a follow-up PR to merge these two Dockerfiles and update the build-release.yaml? The Dockerfile.arm64v8-buildx should be built using `docker buildx build --platform linux/arm64`. Therefore, the build-release.yaml file needs some revamp, and I am currently uncertain about the syntax and how we can test it.

Essentially, the current `docker build` command in build-release.yaml needs to be replaced by `docker buildx build`, and it needs a new flag --platform with the correct values (arm64, amd64, arm32 as opposed to the current arch names arm64v8, arm32v7, x86_64).

niedbalski replied:
@sarandia OK, lets stick this PR to the proposed Cmake changes, which overall seems correct. Please remove the proposed addition of dockerfiles/Dockerfile.arm64v8-dockerx and lets start a new PR to address the switch to buildx independently.

sarandia replied:
OK, done, removed this Dockerfile, and will add you to a new PR for buildx changes.

niedbalski replied:
Great. Please update the commit message and pull request description accordingly.

sarandia replied:
Updated.

sarandia commented on CMakeLists.txt:
```diff
@@ -610,9 +610,9 @@ if(FLB_JEMALLOC AND ${CMAKE_SYSTEM_NAME} MATCHES "Linux")
   ExternalProject_Add(jemalloc
     SOURCE_DIR ${CMAKE_CURRENT_SOURCE_DIR}/lib/jemalloc-5.2.1
-    CONFIGURE_COMMAND ${CMAKE_CURRENT_SOURCE_DIR}/lib/jemalloc-5.2.1/configure ${AUTOCONF_HOST_OPT} --with-lg-quantum=3 --prefix=<INSTALL_DIR>
+    CONFIGURE_COMMAND ${CMAKE_CURRENT_SOURCE_DIR}/lib/jemalloc-5.2.1/configure ${AUTOCONF_HOST_OPT} --host=${HOST} --with-lg-quantum=3 --prefix=<INSTALL_DIR>
```
https://github.com/fluent/fluent-bit/blob/master/lib/jemalloc-5.2.1/configure#L1426

The configure script of jemalloc (and onigmo) use this variable to cross-compile for a certain paltform.

niedbalski reviewed (approved):
LGTM

niedbalski commented:
@sarandia please check the failed commit message check. https://github.com/fluent/fluent-bit/pull/3152/checks?check_run_id=2083040159

sarandia commented:
> @sarandia please check the failed commit message check. https://github.com/fluent/fluent-bit/pull/3152/checks?check_run_id=2083040159

Fixed (and updated the PR title to be in sync)

niedbalski commented:
@edsiper I am +1 on this, please review for a +2.

edsiper commented:
thanks!
